File: packages/ipywxyz/src/ipywxyz/widgets/jsonld.py
""" Widgets for working with JSON
"""
from pyld import jsonld

from .base import Fn, T, W


@W.register
class Expand(Fn):
    """ Expand a JSON document to a list of nodes
    """

    _model_name = T.Unicode("ExpandModel").tag(sync=True)

    source = T.Dict(allow_none=True).tag(sync=True)
    value = T.List(allow_none=True).tag(sync=True)
    context = T.Dict(allow_none=True).tag(sync=True)

    @T.default("source")
    @T.default("context")
    def _default_source_contex(self):
        return {}

    @T.default("value")
    def _default_value(self):
        return []

    @T.observe("source", "context")
    def _change(self, *_):
        try:
            self.error = ""
            self.value = jsonld.expand(self.source, dict(
                expandContext=self.context
            ))
        except Exception as err:
            self.error = f"{err}"


@W.register
class Compact(Fn):
    """ Compact a JSON document with a context
    """

    _model_name = T.Unicode("CompactModel").tag(sync=True)

    source = T.Dict(allow_none=True).tag(sync=True)
    value = T.Dict(allow_none=True).tag(sync=True)
    context = T.Dict(allow_none=True).tag(sync=True)
    expand_context = T.Dict(allow_none=True).tag(sync=True)

    @T.default("source")
    @T.default("context")
    @T.default("value")
    @T.default("expand_context")
    def _default(self):
        return {}

    @T.observe("source", "context", "expand_context")
    def _change(self, *_):
        try:
            self.error = ""
            self.value = jsonld.compact(self.source, self.context, dict(
                expandContext=self.expand_context
            ))
        except Exception as err:
            self.error = f"{err}"


@W.register
class Flatten(Fn):
    """ Flatten a JSON document to a flat graph
    """

    _model_name = T.Unicode("FlattenModel").tag(sync=True)

    source = T.Dict(allow_none=True).tag(sync=True)
    value = T.Dict(allow_none=True).tag(sync=True)
    context = T.Dict(allow_none=True).tag(sync=True)
    expand_context = T.Dict(allow_none=True).tag(sync=True)

    @T.default("source")
    @T.default("context")
    @T.default("value")
    @T.default("expand_context")
    def _default(self):
        return {}

    @T.observe("source", "context", "expand_context")
    def _change(self, *_):
        try:
            self.error = ""
            self.value = jsonld.flatten(self.source, self.context, dict(
                expandContext=self.expand_context
            ))
        except Exception as err:
            self.error = f"{err}"


@W.register
class Frame(Fn):
    """ Frame a JSON document
    """

    _model_name = T.Unicode("FrameModel").tag(sync=True)

    source = T.Dict(allow_none=True).tag(sync=True)
    value = T.Dict(allow_none=True).tag(sync=True)
    frame = T.Dict(allow_none=True).tag(sync=True)
    expand_context = T.Dict(allow_none=True).tag(sync=True)

    @T.default("source")
    @T.default("frame")
    @T.default("expand_context")
    @T.default("value")
    def _default(self):
        return {}

    @T.observe("source", "frame", "expand_context")
    def _change(self, *_):
        try:
            self.error = ""
            self.value = jsonld.frame(self.source, self.frame, dict(
                expandContext=self.expand_context
            ))
        except Exception as err:
            self.error = f"{err}"


@W.register
class Normalize(Fn):
    """ Normalize a JSON document
    """

    _model_name = T.Unicode("NormalizeModel").tag(sync=True)

    source = T.Dict(allow_none=True).tag(sync=True)
    value = T.Union([
        T.Dict(allow_none=True),
        T.Unicode(allow_none=True)
    ]).tag(sync=True)
    expand_context = T.Dict(allow_none=True).tag(sync=True)
    format = T.Unicode(default_value="application/n-quads",
                       allow_none=True).tag(sync=True)

    @T.default("source")
    @T.default("expand_context")
    @T.default("value")
    def _default(self):
        return {}

    @T.observe("source", "expand_context", "format")
    def _change(self, *_):
        try:
            self.error = ""
            self.value = jsonld.normalize(self.source, dict(
                expandContext=self.expand_context,
                format=self.format
            ))
        except Exception as err:
            self.error = f"{err}"

File: src/ts/wxyz-jsonld/src/widgets/jsonld.ts
// this is a name-only import
import * as jsonld from 'jsonld';

import { JSONLDBase } from './_jsonld';

import { TObject } from '@deathbeds/wxyz-core/lib/widgets/_base';

export class ExpandModel extends JSONLDBase<
  TObject,
  TObject[],
  ExpandModel.ITraits
> {
  static model_name = 'ExpandModel';

  defaults() {
    return { ...super.defaults(), _model_name: ExpandModel.model_name };
  }

  async theFunction(source: TObject) {
    const context = this.theExpandContext;
    const opts = context ? { expandContext: context } : {};
    return await jsonld.expand(source, opts);
  }
}

export class CompactModel extends JSONLDBase<
  TObject,
  TObject,
  CompactModel.ITraits
> {
  static model_name = 'CompactModel';

  defaults() {
    return { ...super.defaults(), _model_name: CompactModel.model_name };
  }

  initialize(attributes: any, options: any) {
    super.initialize(attributes, options);
    this.on('change:context', this.theSourceChanged, this);
    return this;
  }

  async theFunction(source: TObject) {
    const expandContext = this.theExpandContext;
    const context = this.get('context') as jsonld.IContext;
    return await jsonld.compact(source, context, {
      ...(expandContext ? { expandContext } : {}),
    });
  }
}

export class FlattenModel extends JSONLDBase<
  TObject,
  TObject,
  FlattenModel.ITraits
> {
  static model_name = 'FlattenModel';

  defaults() {
    return { ...super.defaults(), _model_name: FlattenModel.model_name };
  }

  initialize(attributes: any, options: any) {
    super.initialize(attributes, options);
    this.on('change:context', this.theSourceChanged, this);
    return this;
  }

  async theFunction(source: TObject) {
    const expandContext = this.theExpandContext;
    const context = this.get('context') as jsonld.IContext;
    return await jsonld.flatten(source, context, {
      ...(expandContext ? { expandContext } : {}),
    });
  }
}

export class FrameModel extends JSONLDBase<
  TObject,
  TObject,
  FrameModel.ITraits
> {
  static model_name = 'FrameModel';

  defaults() {
    return { ...super.defaults(), _model_name: FrameModel.model_name };
  }

  initialize(attributes: any, options: any) {
    super.initialize(attributes, options);
    this.on('change:frame', this.theSourceChanged, this);
    return this;
  }

  async theFunction(source: TObject) {
    const expandContext = this.theExpandContext;
    const frameContext = this.get('frame') as jsonld.IContext;
    return await jsonld.frame(source, frameContext, {
      ...(expandContext ? { expandContext } : {}),
    });
  }
}

export class NormalizeModel extends JSONLDBase<
  TObject,
  TObject | string,
  NormalizeModel.ITraits
> {
  static model_name = 'NormalizeModel';

  defaults() {
    return { ...super.defaults(), _model_name: NormalizeModel.model_name };
  }

  initialize(attributes: any, options: any) {
    super.initialize(attributes, options);
    this.on('change:format', this.theSourceChanged, this);
    return this;
  }

  async theFunction(source: TObject) {
    const expandContext = this.theExpandContext;
    const format = this.get('format');
    return await jsonld.normalize(source, {
      ...(expandContext ? { expandContext } : {}),
      ...(format ? { format } : {}),
    });
  }
}

export namespace ExpandModel {
  export interface ITraits<V = TObject, W = TObject[]>
    extends JSONLDBase.ITraits<V, W> {
    expand_context: TObject;
  }
}

export namespace CompactModel {
  export interface ITraits<V = TObject, W = TObject>
    extends JSONLDBase.ITraits<V, W> {
    context: jsonld.IContext;
  }
}

export namespace FlattenModel {
  export interface ITraits<V = TObject, W = TObject>
    extends JSONLDBase.ITraits<V, W> {
    context: jsonld.IContext;
  }
}

export namespace FrameModel {
  export interface ITraits<V = TObject, W = TObject>
    extends JSONLDBase.ITraits<V, W> {
    frame: jsonld.IContext;
  }
}

export namespace NormalizeModel {
  export interface ITraits<V = TObject, W = TObject | string>
    extends JSONLDBase.ITraits<V, W> {
    format: string;
  }
}
